### DISCUSSION TITLE
Understanding the source code of jax.lax | tanh

### DISCUSSION THREAD

**Discussion by Rohanjames1997** [jax-ml/jax#24240] [Q&A] [answered]
Hi there,

I'm trying to understand how the `tanh` function is implemented in jax, and I'm interested in knowing where the actual underlying "math" resides in the codebase. 

Does it simply call numpy's/XLA's tanh operation? If I were to optimize the `tanh` op in jax using SIMD intrinsics, where should I be looking to do so?

So far, from the repo, I've gotten this, but I couldn't make much of it : https://github.com/jax-ml/jax/blob/8ef41a6e14e6620426791029919bb691d4325806/jax/_src/lax/lax.py#L337-L339

Any pointers appreciated, TIA!

**Comment by dfm** [comment, marked as answer]
It's not so straightforward to follow the full stack, but the basic flow is that a JAX function get's lowered to [StableHLO](https://github.com/openxla/stablehlo), which you can inspect as follows:

```python
import jax
import jax.numpy as jnp

print(jax.jit(jnp.tanh).lower(jnp.linspace(-1, 1, 5)).as_text())
```

Since [StableHLO has a tanh intrinsic](https://openxla.org/stablehlo/spec#tanh), the intermediate representation is (relatively) simple:

```
module @jit_tanh attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<5xf32> {mhlo.layout_mode = "default"}) -> (tensor<5xf32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.tanh %arg0 : tensor<5xf32>
    return %0 : tensor<5xf32>
  }
}
```

Then, depending on the specific platform that you are targeting, [XLA](https://github.com/openxla/xla) will compile this program differently. For example, on CPU, I believe the tanh will be passed to LLVM, which (I expect) will emit SIMD operations.

So, the tl;dr is that it's not really straightforward to track down exactly how specific operations are compiled. If you want to contribute optimizations like this, those would live in LLVM or XLA rather than JAX itself.

Hope this helps!

**Reply by Rohanjames1997** [reply]
Thank you - this helps a lot!

Indeed - while debugging I found that on a CPU, the default XLA implementation of float32 tanh on aarch64 is in [EmitFastTanh](https://github.com/openxla/xla/blob/53044afbc3f5fe84f1bb1ee5aebbe7100ce57132/xla/service/llvm_ir/math_ops.cc#L30-L86) which in-turn emits LLVM SIMD ops like you said.
